packages/front-end/components/Experiment/EditDOMMutationsModal.tsx

```diff
@@ -65,67 +81,82 @@ const EditDOMMutatonsModal: FC<{
-                  delete
+                  clear
                 </a>
               </small>
             ) : null}
           </h4>
-          {newVisualChange.css ? (
-            <Code
-              language="css"
-              code={newVisualChange.css}
-              className="disabled"
-            />
-          ) : (
-            <div className="text-muted font-italic">(None)</div>
-          )}
+
+          <Field
+            textarea
+            minRows={5}
+            value={css}
+            onChange={(e) => {
+              setCss(e.target.value);
+              setNewVisualChange({
+                ...newVisualChange,
+                css: e.target.value,
+              });
+            }}
+          />
         </div>
 
         <div className="mb-4">
           <h4>
             Custom JS
             {newVisualChange.js ? (
-              <small className="ml-2">
+              <small className="ml-2 float-right">
                 <a href="#" className="text-danger" onClick={deleteCustomJS}>
-                  delete
+                  clear
                 </a>
               </small>
             ) : null}
           </h4>
-          {newVisualChange.js ? (
-            <Code
-              language="javascript"
-              code={newVisualChange.js ?? ""}
-              className="disabled"
-            />
-          ) : (
-            <div className="text-muted font-italic">(None)</div>
-          )}
+
+          <Field
+            textarea
+            minRows={5}
+            value={js}
+            onChange={(e) => {
+              setJs(e.target.value);
+              setNewVisualChange({
+                ...newVisualChange,
+                js: e.target.value,
+              });
+            }}
+          />
         </div>
 
         <div className="mb-4">
           <h4>DOM Mutations</h4>
-
           {newVisualChange.domMutations.length ? (
             newVisualChange.domMutations.map((m, i) => (
-              <div key={i} className="d-flex flex-column align-items-end">
+              <Fragment key={i}>
                 <a
-                  className="text-danger"
+                  className="text-danger float-right"
                   href="#"
                   onClick={() => deleteDOMMutation(i)}
                   style={{ marginBottom: "-.5rem", fontSize: "0.75rem" }}
                 >
                   delete
                 </a>
-                <Code
-                  language="json"
-                  code={JSON.stringify(m)}
-                  className="disabled"
-                  containerClassName="w-100"
+                <Field
+                  textarea
+                  minRows={2}
+                  value={JSON.stringify(m)}
+                  className="w-100 my-3"
+                  onChange={(e) => {
+                    try {
+                      const newMutation = JSON.parse(e.target.value);
+                      setDOMMutation(i, newMutation);
```

Comment:
We need more validation here.  Otherwise, someone could cause some serious bugs in the SDK
1. All required fields are set
2. There are no extra fields
3. The `action` is a valid value

Here's a type to work from for validation: https://github.com/growthbook/dom-mutator/blob/main/src/index.ts#L515

Reply:
Done